nixos/modules/virtualisation/digital-ocean-config.nix

```diff
@@ -0,0 +1,163 @@
+{ config, pkgs, lib, modulesPath, ... }:
+with lib;
+{
+  imports = [
+    (modulesPath + "/profiles/qemu-guest.nix")
+    (modulesPath + "/virtualisation/digital-ocean-init.nix")
+  ];
+  options.virtualisation.digitalOcean = with types; {
+    setRootPassword = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to set the root password from the Digital Ocean metadata";
+    };
+    setSshKeys = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to fetch ssh keys from Digital Ocean";
+    };
+    seedEntropy = mkOption {
+      type = bool;
+      default = true;
+      example = true;
+      description = "Whether to run the kernel RNG entropy seeding script from the Digital Ocean vendor data";
+    };
+  };
+  config =
+    let
+      cfg = config.virtualisation.digitalOcean;
+      hostName = config.networking.hostName;
+    in mkMerge [{
+      fileSystems."/" = {
+        device = "/dev/disk/by-label/nixos";
+        autoResize = true;
+        fsType = "ext4";
+      };
+      boot = {
+        growPartition = true;
+        kernelParams = [ "console=ttyS0" "panic=1" "boot.panic_on_fail" ];
+        initrd.kernelModules = [ "virtio_scsi" ];
+        kernelModules = [ "virtio_pci" "virtio_net" ];
+        loader = {
+          grub.device = "/dev/vda";
+          timeout = 0;
+          grub.configurationLimit = 0;
+        };
+      };
+      services.openssh = {
+        enable = mkDefault true;
+        passwordAuthentication = mkDefault false;
+      };
+      networking = {
+        firewall.allowedTCPPorts = [ 22 ];
+        hostName = mkDefault ""; # use Digital Ocean metadata server
+        useNetworkd = mkDefault true;
```

Comment:
Why have `networkd` by default? This is an experimental feature that still has some known bugs.  I'd prefer not enabling it by default on installations. Especially because it behaves differently from what people expect

Reply:
Turns out the answer is: Because I added it there. Which was a stupid mistake on my side! :D. we should remove it

Reply:
Hm, it's been working well for me... From initial testing turning it off it causes problems for the metadata init services and neither hostname nor ssh keys are making it to the VM :(

Reply:
That sounds like a (serious) bug in the NixOS networking module.  By default `networking` module should listen to DHCP requests on all interfaces. investigating...

Reply:
New theory `network-online.target` works well on `systemd-networkd` but maybe not with the default networking module.    Workaround: add a `Restart=always` to `digitalocean-init` such that it keeps restarting until the network is available instead of it failing immediately...

This is me purely speculating though; I'll spin up a digitalocean box and do some actual debugging later tonight

Reply:
Yeah, currently looking into it a bit... I think the `network-online.target` is fine, the problem is that [I'm aiming to have the services access the link-local metadata address before the network comes online](https://github.com/NixOS/nixpkgs/commit/7a831ae66616dee57f714fe5618dd8802dfb1400) (start after `network.target`), because the hostname in particular gets set too late otherwise. My assumptions there must be wrong in some way, need to look into it while still having access to the machine ><

Reply:
I'd like it to work before network-online, because default dhcpcd is very slow so delays boot and settings (the hostname being set late causes some services to just call the machine "localhost", and entropy being set late kind of sucks for sshd keygen), but...

1. Any approach that sets the link-local address on the interface before network setup could work but seems fragile?
2. Though systemd-networkd works much better, it seems to only be happenstance that `network.target` works just because it's so fast.
3. curl --retry options don't help because it checks for specific errors that curl considers transient. Connection failures happen to not be one of those!
4. systemd's `Restart=` would help if the services are changed to `Type=simple`, but would have to decide on throttle/interval settings. This seems like the best approach for making it tolerant to failure?

For now, going back to `After=network-online.target` but saving to `/etc/hostname` (if not otherwise set by nixos config) so that at least it will be available on next reboot.

Reply:
How about we just slam the metadata unit in a bash while loop?

Reply:
You could  even make the unit part of `sysinit.target` if you're careful. It will then do entropy initialisation before any other services are started.

Reply:
> How about we just slam the metadata unit in a bash while loop?

Heh that's an option too! Though...
1. What retry logic to use? Sleep between attempts? Exponential backoff? Throttle or timeout after so many attempts or time has passed?
2. There are multiple commands that need this, so I guess basically it'd be a wrapper script around curl that the units call into instead of a loop around the unit.
    - Hm, alternatively you could maybe make one unit fetch metadata and trigger the others once it's succeeded, either via requires+before or via a path unit? It seems like a good idea to have one unit serve as a "metadata accessible" dependency marker and let the others just wait on it.

> You could even make the unit part of sysinit.target if you're careful. It will then do entropy initialisation before any other services are started.

At that point I wonder if we should just set up the link-local address then? If it's started early with no other changes it will just stall until the network routes are set up via dhcp, and not really run any earlier than it does otherwise.

Reply:
I just discovered this package by coreos called `afterburn` which implements this step of boot up on clouds in a generic way. Perhaps we could use it to unify all our cloud image metadata fetching steps? (Just a thought; and probably for a later PR). Anyhow, their tactic is just to retry  with an exponetnial backoff  until the link local address is there https://github.com/coreos/afterburn/blob/master/src/providers/digitalocean/mod.rs#L70-L81 https://github.com/coreos/afterburn/blob/861fe5b7203c1a9528bfe8d309959eab9bec07c2/src/retry/mod.rs

Reply:
@arcnmx  I think we can very quickly get this to work by writing
`curl --retry-connrefused` instead of `curl` in all the `curl` calls. I just did this on my digitalocean image and it seems to work fine!

Reply:
> I just did this on my digitalocean image and it seems to work fine! curl will then do the retries with exponential backoff and whatnot

Weird, depending on which network target (link a diff if possible)? I tried connrefused but it did not work for me because [curl seems to give up if the error specifically isn't ECONNREFUSED](https://github.com/curl/curl/issues/1603), which is generally the case when the network isn't available or the metadata server isn't reachable/routable.

> Perhaps we could use it to unify all our cloud image metadata fetching steps? (Just a thought; and probably for a later PR).

Definitely worth looking into! But yeah, not necessarily for this PR.

> They have a service called coreos-metadata.service

Cool, yeah, I think in general that's the way forward here. Whether built on top of afterburn or a bash loop around curl or anything else, it gives us a central point to place the init/detection logic. I'll at least restructure the units around this approach, then even if we kept the network-online dependency for now, it should be easy to make follow-up improvements. I'll also look into writing a retry loop, I'm just slightly wary about making sure that I choose the appropriate retry logic.

Reply:
Oh I'm running this on `networkd` I just realised.  That's probably why it works.

If you're wary about the retry-loop, we can also let systemd handle that:
```
[Service]
Restart=on-failure
RestartSec=zzz
StartLimitIntervalSec=yyy
StartLimitBurst=xxx
```

I think either way works. Systemd doesn't do exponential backoff though, but im not sure we require that. the link-local server isn't gonna crash from hammering.  Restarting every 2 seconds for like 10 times is probably sufficient.

I'll see if I can package afterburner tonight

Reply:
> If you're wary about the retry-loop, we can also let systemd handle that, which is maybe more elegant.

Though I wanted to, it requires that the service type be changed to `simple`, which then makes it unusable as a dependency for other units. It does work okay for the individual units, but yeah the lack of backoff is slightly annoying... A loop will work fine.

> the link-local server isn't gonna crash from hammering. Restarting every 2 seconds for like 10 times is probably sufficient

Probably, my concern is more that we don't want it to spam journald forever, but we do want to make sure it doesn't give up too early if the network is slow to start or whatever, because these services are important for making sure you can access the machine! While also being fairly responsive because if you wait 2 seconds, a bunch of services will start in the meantime.

Reply:
yeh not having it a oneshot is annoying. I didn't realise `oneshot` didn't support `Restart=`

Reply:
So, pushed a few things:
1. metadata.service added with a retry loop that can be used as a dependency so that services that need metadata can just wait on it before doing their thing. The service is ordered After dhcpcd or systemd-networkd (if you have them enabled) and thus the retry loop isn't likely to be needed except in unusual circumstances, as the service should just start as soon as an ip/route is assigned to the interface.
2. sethostname and entropy services are wantedBy and Before network.target, so that any services that are ordered After network or network-online will wait for them. This does have the effect of delaying the startup of certain services until the droplet has an ip, but that's also kind of the point. networkd is pretty instantaneous, dhcpcd takes about 6 seconds.
3. Some fixes to the userdata init service. It remains part of / After network-online.
4. the password/sshkey services are just part of mult-user but ordered to take place before sshd so you can't have racy scenarios where the server is online and reachable but are unable to login.
5. Some changes to image/init to allow them to share the config file because it seemed kind of silly to duplicate?